### ISSUE TITLE
Cannot disable GPU using conda environment

### ISSUE THREAD

**Issue by eafpres** [rstudio/keras#764]
I tried the solution to this closed issue:
https://github.com/rstudio/keras/issues/406

`# Disable GPU                                                                   
Sys.setenv("CUDA_VISIBLE_DEVICES" = -1)   `

but then when using the following, the gpu is still visible to keras and is used:
`  reticulate::use_condaenv("keras-gpu")
  library(keras)`

I imagine it is because of explicitly loading conda env "keras-gpu"
Do I need to create a conda env without gpu and use that, or is there another way?

**Comment by skeydan** [comment]
Hi, 

I always create different envs but now I actually tried what you say

```
Sys.setenv("CUDA_VISIBLE_DEVICES" = -1)

reticulate::use_condaenv("tf-0413-gpu", required = T)
library(keras)
x = 1.1:10000.1 %>% as.matrix()
y=1:10000 %>% as.matrix()
model <- keras_model_sequential()
model %>% layer_dense(units=1)
model %>% compile(optimizer="adam", loss="mse")
model %>% fit(x , y, epochs = 100)
```

and it does _not_ use the GPU, for me.
The gpu-related logging output is

```
2019-04-29 08:39:33.294045: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcuda.so.1
2019-04-29 08:39:33.297871: E tensorflow/stream_executor/cuda/cuda_driver.cc:320] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2019-04-29 08:39:33.297900: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:166] retrieving CUDA diagnostic information for host: snowflake
2019-04-29 08:39:33.297906: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:173] hostname: snowflake
2019-04-29 08:39:33.297957: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:197] libcuda reported version is: 418.56.0
2019-04-29 08:39:33.297976: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:201] kernel reported version is: 418.56.0
2019-04-29 08:39:33.297981: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:308] kernel version seems to match DSO: 418.56.0
```

Note the line 

```
CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
```

**Comment by eafpres** [comment]
Thank you, I will rerun it and provide output.

**Comment by eafpres** [comment]
I retested this and it does appear to disable the GPU.

`Sys.setenv("CUDA_VISIBLE_DEVICES" = -1)
reticulate::use_condaenv("keras-gpu")
library(keras)
use_session_with_seed(seed, disable_gpu = FALSE, disable_parallel_cpu = FALSE)`

2019-05-01 20:23:50.561212: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX AVX2
2019-05-01 20:23:50.935424: E tensorflow/stream_executor/cuda/cuda_driver.cc:300] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2019-05-01 20:23:50.966298: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:161] retrieving CUDA diagnostic information for host: EAF-LLC
2019-05-01 20:23:50.967514: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:168] hostname: EAF-LLC
Set session seed to 20190424

Interestingly, this does not work:
`reticulate::use_condaenv("keras-gpu")
library(keras)
use_session_with_seed(seed, disable_gpu = TRUE, disable_parallel_cpu = FALSE)`

2019-05-01 20:26:22.402708: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX AVX2
2019-05-01 20:26:22.902278: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1433] Found device 0 with properties: 
name: GeForce GTX 860M major: 5 minor: 0 memoryClockRate(GHz): 1.0195
pciBusID: 0000:01:00.0
totalMemory: 2.00GiB freeMemory: 1.64GiB
2019-05-01 20:26:22.903163: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1512] Adding visible gpu devices: 0
2019-05-01 20:26:24.743460: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-05-01 20:26:24.744157: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990]      0 
2019-05-01 20:26:24.744574: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1003] 0:   N 
2019-05-01 20:26:24.745427: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 1387 MB memory) -> physical GPU (device: 0, name: GeForce GTX 860M, pci bus id: 0000:01:00.0, compute capability: 5.0)
Set session seed to 20190424 (disabled GPU)

`  train_history_subset_1 <-
    model %>%
    fit_generator(
      data_gen,
      steps_per_epoch = steps_per_epoch_train,
      validation_data = val_gen,
      validation_steps = steps_per_epoch_val,
      epochs = 600)`

2019-05-01 20:28:01.836117: I tensorflow/stream_executor/dso_loader.cc:152] successfully opened CUDA library cublas64_100.dll locally

And I verified by timing it is definitely using the GPU.  So it sees it, says it is disabled, but then uses it anyway.  Perhaps I'm using the use_session_with_seed() incorrectly.

Regardless, it does work the way I originally posted, so I'm closing this issue.
